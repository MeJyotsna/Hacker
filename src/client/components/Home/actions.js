import "regenerator-runtime/runtime";

const BASE_URL = "https://hn.algolia.com/api/v1";
const FETCH_NEWS_FEED = "FETCH_NEWS_FEED";
const SET_UPVOTE_COUNT = "SET_UPVOTE_COUNT";
const NEWS_STORAGE_KEY = "news_feed";
const NEWS_FEED_SHOW = 0;
const NEWS_FEED_HIDE = 1;

export const dispatchNewsFeeds = (data) => ({
  type: FETCH_NEWS_FEED,
  payload: data,
});
// export const fetchNewsFeedTest = async (base, destination) => {
//   try {
//     const result = await fetch(
//       `https://api.exchangeratesapi.io/latest?base=${base}`
//     );
//     const data = await result.json();
//     console.log("------------", data);
//     return data.rates[destination];
//   } catch (e) {
//     return null;
//   }
// };
export const fetchNewsFeed = (start, end) => async (dispatch) => {
  let feeds = [];
  for (let i = start; i < end; i++) {
    await fetch(BASE_URL + "/items/" + parseInt(i))
      .then((response) => response.json())
      .then((data) => {
        let results = data;
        if (results.title != null) {
          console.log("----from actions page api call result", results);
          let title = results.title;
          let id = results.id;
          let author = results.author;
          let url = results.url;
          let timeStamp = results.created_at_i;
          let vote_count = results.points;
          let commentCount = results.children.length;
          let date = new Date(timeStamp * 1000);
          let posted_time =
            date.getMonth() + "/" + date.getDay() + "/" + date.getFullYear();
          let storage_item = localStorage.getItem(NEWS_STORAGE_KEY + id);
          let parse_storage_item = JSON.parse(storage_item);
          let vote_storage_count =
            parse_storage_item != null ? parse_storage_item.vote_count : 0;
          let show_hide_status =
            parse_storage_item != null &&
            parse_storage_item.hide &&
            parse_storage_item.hide === NEWS_FEED_HIDE
              ? NEWS_FEED_HIDE
              : NEWS_FEED_SHOW;
          if (show_hide_status === NEWS_FEED_SHOW) {
            let news_results = {
              id: id,
              title: title,
              author: author,
              time: posted_time,
              url: url,
              vote_count:
                vote_storage_count > vote_count
                  ? vote_storage_count
                  : vote_count,
              comments: commentCount,
              hide: show_hide_status,
            };
            feeds.push(news_results);
            localStorage.setItem(
              NEWS_STORAGE_KEY + id,
              JSON.stringify(news_results)
            );
            dispatch(dispatchNewsFeeds(feeds));
          }
        }
      })
      .catch((error) => {
        console.log("Error Feed -- " + i + " == " + error);
      });
  }
};

export const dispatchVoteCount = (count) => ({
  type: SET_UPVOTE_COUNT,
  payload: count,
});

export const setUpVoteCount = (vote) => (dispatch) => {
  console.log("Vote : setUpVoteCount Redux = " + JSON.stringify(vote));
  dispatch(dispatchVoteCount(vote));
};

export const hideNewsFeed = (feeds, id) => (dispatch) => {
  if (feeds != null) {
    let parseHide = JSON.parse(JSON.stringify(feeds));
    if (parseHide !== undefined) {
      let news = parseHide;
      let index = news.findIndex((el) => el.id === id);
      let item = news[index];
      news.splice(index, 1);
      let hide_results = {
        id: item.id,
        title: item.title,
        author: item.author,
        time: item.posted_time,
        url: item.url,
        vote_count: item.vote_count,
        comments: item.comments,
        hide: NEWS_FEED_HIDE,
      };
      localStorage.setItem(NEWS_STORAGE_KEY + id, JSON.stringify(hide_results));
      dispatch(dispatchNewsFeeds(news));
    }
  }
};
