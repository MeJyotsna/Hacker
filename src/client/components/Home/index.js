import React from "react";
import { connect } from "react-redux";
import { fetchNewsFeed, setUpVoteCount, hideNewsFeed } from "./actions";
import { LineChart } from "react-chartkick";
import "chart.js";
import * as moment from "moment";

class HomeComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      pageNumber: 0,
    };
  }

  componentDidMount() {
    this.props.fetchNewsFeed(this.state.pageNumber);
  }

  // to open last page visited
  previousPage() {
    let page_number = this.state.pageNumber;
    if (page_number > 0) {
      let count = page_number - 1;
      this.setState({
        pageNumber: count,
      });
      this.props.fetchNewsFeed(count);
    }
  }

  //to open next page
  nextPage() {
    let page_number = this.state.pageNumber;
    let count = page_number + 1;
    this.setState({
      pageNumber: count,
    });
    this.props.fetchNewsFeed(count);
  }

  //rendering chart with the available page data
  renderChart = (results) => {
    var statistics = [];
    for (var i = 0; i < results.length; i++) {
      var id = results[i].id.toString();
      var storage_item = localStorage.getItem("news_feed" + id);
      var parse_storage_item = JSON.parse(storage_item);
      var vote_count =
        parse_storage_item != null ? parse_storage_item.vote_count : 0;
      var votes = vote_count;
      var item = { [id]: votes };
      statistics.push(item);
    }

    //stringifying and data handling as per react-chartkick
    var graph_item = JSON.stringify(statistics);
    graph_item = graph_item.replace(/[{}]/g, "");
    graph_item = graph_item.replace("[", "{");
    graph_item = graph_item.replace("]", "}");
    var data = JSON.parse(graph_item);
    return <LineChart data={data} />;
  };

  //open link in new tab
  getUrl(el) {
    if (el != undefined) {
      var ol = el
        .replace("http://", "")
        .replace("https://", "")
        .split(/[/?#]/)[0];
      return ol;
    }
  }

  //rendering latest vote count as per user events
  renderVoteCount = (item) => {
    var id = item.id;
    var votes;
    var vote = JSON.parse(JSON.stringify(this.props.voteCount)).voteCount;
    if (id == vote.id) {
      votes = vote.vote_count;
    } else {
      var storage_item = localStorage.getItem("news_feed" + id);
      var parse_storage_item = JSON.parse(storage_item);
      var voteCount =
        parse_storage_item != null ? parse_storage_item.vote_count : 0;
      votes = voteCount;
    }
    return <span className="">{votes}</span>;
  };

  handleUpvote = (item) => {
    var id = item.id;
    var title = item.title;

    //fetching news feed with clicked news id
    var storageItem = localStorage.getItem("news_feed" + id);
    var parsedStorageItem = JSON.parse(storageItem);
    var parsedStorageTitle = parsedStorageItem.title;
    var parsedStorageVote = parsedStorageItem.vote_count;

    //check if the clicked data matches local storage's news feed
    if (title == parsedStorageTitle) {
      var newVoteCount = parseInt(parsedStorageVote) + 1;
      var news_results = item;
      news_results.showUpvoteBtn = false;
      news_results.vote_count = newVoteCount;
      localStorage.setItem("news_feed" + id, JSON.stringify(news_results));
      this.props.setUpVoteCount(news_results);
    }
  };

  // to hide the specific news feed
  handleNewsHide = (newsId) => {
    this.props.hideNewsFeed(this.props.newsFeed.newsFeed, newsId);
  };

  render() {
    const listData = JSON.parse(JSON.stringify(this.props.newsFeed.newsFeed)); // getting newsfeed data from props
    return (
      <div className="container">
        <table
          className="table"
          cellPadding="0"
          cellSpacing="0"
          width="100%"
          border="0"
        >
          <thead>
            <tr>
              <th>Comments</th>
              <th>Vote Count</th>
              <th>Upvote</th>
              <th>News details</th>
            </tr>
          </thead>
          <tbody>
            {listData && listData.length === 0 && (
              <tr>
                <td>No Data Found.</td>
              </tr>
            )}
            {listData &&
              listData.length > 0 &&
              listData.map((item) => {
                return (
                  <tr>
                    <td>{item.comments}</td>
                    <td>{this.renderVoteCount(item)}</td>
                    <td>
                      <img
                        onClick={() => this.handleUpvote(item)}
                        className="upvotebtn"
                        src="https://news.ycombinator.com/grayarrow2x.gif"
                        alt="upvote"
                        title="upvote"
                      />
                    </td>
                    <td>
                      {item.title}{" "}
                      <span>
                        <a
                          href={item.url}
                          className=""
                          target="_blank"
                          className="linkColor"
                        >
                          ({this.getUrl(item.url)})
                        </a>{" "}
                        <span className="linkColor">by</span>{" "}
                        <b>{item.author}</b>
                        <span className="linkColor">
                          {" "}
                          {moment(new Date(item.time), "").fromNow()}
                        </span>
                        <span
                          className="hideBtn linkColor"
                          onClick={() => this.handleNewsHide(item.id)}
                        >
                          {" "}
                          [ <span className="btn_hide">hide</span>]
                        </span>
                      </span>
                    </td>
                  </tr>
                );
              })}
          </tbody>
        </table>
        <div className="clearfix">
          <div className="float-right">
            <div className="prevBtn btn" onClick={() => this.previousPage()}>
              Previous
            </div>
            <div className="nextBtn btn" onClick={() => this.nextPage()}>
              Next
            </div>
          </div>
        </div>
        <hr />
        <div className="clearfix footer-wrap">
          <div>
            <p className="votes">Votes</p>
          </div>
          {this.renderChart(listData)}
        </div>
        <div className="id-wrap">
          <p className="id">ID</p>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  newsFeed: state.newsFeed,
  voteCount: state.voteCount,
});

const mapDispatchToProps = (dispatch) => ({
  fetchNewsFeed: (start, end) => dispatch(fetchNewsFeed(start, end)),
  setUpVoteCount: (item) => dispatch(setUpVoteCount(item)),
  hideNewsFeed: (feeds, id) => dispatch(hideNewsFeed(feeds, id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(HomeComponent);
