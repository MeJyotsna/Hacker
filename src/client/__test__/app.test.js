import React from "react";
import Enzyme, { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Provider } from "react-redux";
import "chart.js";
import store from "../Store";
import HomeComponent from "../components/Home";
import { fetchNewsFeed, dispatchNewsFeeds } from "../components/Home/actions";

Enzyme.configure({ adapter: new Adapter() });
let wrapper;
describe("When home component is mounted", () => {
  beforeEach(() => {
    wrapper = mount(
      <Provider store={store}>
        <HomeComponent />
      </Provider>
    );
  });
  it("should check for table", () => {
    wrapper = mount(
      <Provider store={store}>
        <HomeComponent />
      </Provider>
    );
    expect(wrapper.find("table")).toHaveLength(1);
  });

  it("should check for snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should check for dispatchNewsFeeds function to have 'FETCH_NEWS_FEED'", () => {
    var val = dispatchNewsFeeds({});
    expect(val.type === "FETCH_NEWS_FEED").toBe(true);
  });

  it("should check for action has type and payload property", () => {
    var val = dispatchNewsFeeds({});
    expect(val).toHaveProperty("type");
    expect(val).toHaveProperty("payload");
  });

  it("should check for payload has Id property", () => {
    var val = dispatchNewsFeeds({ id: 1 });
    expect(val.payload.id === 1).toBe(true);
  });
});
