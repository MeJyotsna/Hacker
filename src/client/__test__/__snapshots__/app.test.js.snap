// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When home component is mounted should check for snapshot 1`] = `ReactWrapper {}`;
